File: tools/src/pcr_tools/coordinates.py
"""Typed 0-based half-open coordinates used at scientific boundaries."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

Strand = Literal["+", "-"]


@dataclass(frozen=True, slots=True, order=True)
class BaseIndex:
    """Zero-based base index."""

    value: int

    def __post_init__(self) -> None:
        if self.value < 0:
            raise ValueError("base index cannot be negative")


@dataclass(frozen=True, slots=True, order=True)
class Boundary:
    """Boundary between bases in a half-open coordinate system."""

    value: int

    def __post_init__(self) -> None:
        if self.value < 0:
            raise ValueError("boundary cannot be negative")


@dataclass(frozen=True, slots=True)
class HalfOpenInterval:
    """[start,end) interval; start <= end by construction."""

    start: Boundary
    end: Boundary

    def __post_init__(self) -> None:
        if self.end.value < self.start.value:
            raise ValueError("half-open interval end cannot precede start")

    @property
    def length(self) -> int:
        return self.end.value - self.start.value


@dataclass(frozen=True, slots=True)
class CircularPosition:
    """Normalized zero-based position on a circular reference."""

    value: int
    reference_length: int

    def __post_init__(self) -> None:
        if self.reference_length <= 0:
            raise ValueError("circular reference length must be positive")
        object.__setattr__(self, "value", self.value % self.reference_length)

File: crates/pcr-core/src/coordinates.rs
//! Typed zero-based half-open coordinates.
//!
//! Scientific engines may still deserialize legacy integers at their boundary,
//! but interval arithmetic should use these types so a base index, a boundary,
//! and a circular position are not interchangeable integers.

use serde::{Deserialize, Serialize};

/// Zero-based base index.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
pub struct BaseIndex(pub usize);

/// Boundary between bases in `[start,end)` coordinates.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
pub struct Boundary(pub usize);

/// Valid half-open interval.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct HalfOpenInterval {
    /// Inclusive start boundary.
    pub start: Boundary,
    /// Exclusive end boundary.
    pub end: Boundary,
}

impl HalfOpenInterval {
    /// Construct an interval, rejecting a reversed pair.
    pub fn new(start: Boundary, end: Boundary) -> Option<Self> {
        (start <= end).then_some(Self { start, end })
    }

    /// Number of bases covered.
    #[must_use]
    pub fn len(self) -> usize {
        self.end.0 - self.start.0
    }

    /// Whether the interval is empty.
    #[must_use]
    pub fn is_empty(self) -> bool {
        self.start == self.end
    }
}

/// Reference strand; coordinates still increase on the reference.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Strand {
    /// Reference/forward strand.
    Plus,
    /// Reverse-complement strand.
    Minus,
}

/// Position normalized modulo a circular reference length.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct CircularPosition {
    /// Normalized zero-based position.
    pub value: usize,
    /// Circular reference length.
    pub reference_length: usize,
}

impl CircularPosition {
    /// Normalize a raw position on a non-empty circular reference.
    pub fn new(value: usize, reference_length: usize) -> Option<Self> {
        (reference_length > 0).then_some(Self {
            value: value % reference_length,
            reference_length,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn interval_has_half_open_length() {
        assert_eq!(
            HalfOpenInterval::new(Boundary(3), Boundary(8))
                .unwrap()
                .len(),
            5
        );
        assert!(HalfOpenInterval::new(Boundary(8), Boundary(3)).is_none());
    }

    #[test]
    fn circular_position_normalizes() {
        assert_eq!(CircularPosition::new(12, 10).unwrap().value, 2);
    }
}
